docs/root/intro/arch_overview/http/http_routing.rst

```diff
@@ -192,3 +192,59 @@ upon configuration load and cache the contents.
 
 If **response_headers_to_add** has been set for the Route or the enclosing Virtual Host,
 Envoy will include the specified headers in the direct HTTP response.
+
+Routing Via Generic Matching
+----------------------------
+
+Envoy recently added support for utilzing a :ref:`generic match tree <envoy_v3_api_msg_config>` to
+specify the route table. This is a more expressive matching engine than the original one, allowing
+for sublinear matching on arbitrary headers (unlike the original matching engine which could only
+do this for :authority in some cases).
+
+To use the generic matching tree, specify a ref:`matcher <fds>` on a virtual host with a RouteAction action:
+
+```
+  matcher:
+    "@type": type.googleapis.com/envoy.config.common.matcher.v3.Matcher
+    matcher_tree:
+      input:
+        name: request-headers
+        typed_config:
+          "@type": type.googleapis.com/envoy.type.matcher.v3.HttpRequestHeaderMatchInput
+          header_name: :path
+      exact_match_map:
+        map:
+          "/new_endpoint/foo":
+            action:
+              name: route
+              typed_config:
+                "@type": type.googleapis.com/envoy.config.route.v3.Route
```

Comment:
Broadly agree this structure looks great. One observation I have is that making `Route` the action, rather than `RouteAction` the action is a bit unfortunate, since we have a `RouteMatcher` inside of this `Route`. Ideally it would just be `RouteAction`, but these days we have a bunch of stuff such as per-route filter config, response headers to add etc. at this level (since shared with other actions). The idea of splitting match-action has kind of fallen apart (we were shooting for that in the v2 xDS RDS structure).

So, what is here is fine. Maybe you could actually run the `RouteMatcher` on the `Route` you resolve to after the `Matcher` match, to  be able to support things like path rewrite. It's a bit redundant, but it also is O(1) and might only need to be done when you need the match criteria.

Reply:
I agree that using `Route` here seems a little counter-intuitive, but I also agree that it's totally workable.

Another option would be to copy the common route fields into a `CommonRouteConfig` message, and then have each of the action types (`RouteAction`, `RedirectAction`, et al) include a field of type `CommonRouteConfig`.  That might be a little more in the spirit of the generic matching API, where the matching action really ought to be the route action.